### ISSUE TITLE
Micronaut 3.0.0: @RequestScope throws IllegalMonitorStateException with Redis sessions enabled

### ISSUE THREAD

**Issue by chrisparton1991** [micronaut-projects/micronaut-core#6096]
### Expected Behavior

I tried upgrading my Micronaut 2.4.0 application to Micronaut 3.0.0, but found that my `@RequestScope` beans no longer work.

I created a blank app using Micronaut Launch and was able to narrow the issue down to having Redis enabled for HTTP sessions.

The attached GitHub repo has a minimal reproducible case in `com.example.Application`.

### Actual Behaviour

When hitting an authenticated endpoint with Redis enabled, attempting to access a `@RequestScope` bean throws an exception:
```
java.lang.IllegalMonitorStateException: attempt to unlock read lock, not locked by current thread
```

Not sure if related, but I've noticed that with Redis session enabled, my endpoint methods run in the `lettuce-nioEventLoop` thread pool (e.g. `Thread[lettuce-nioEventLoop-4-2,5,main]`). With redis sessions disabled, the endpoint runs in the `default-nioEventLoopGroup` pool. If I use `@ExecuteOn(TaskExecutors.IO)`, the endpoint runs in the `io-executor-thread` pool as expected.

### Steps To Reproduce

Clone https://github.com/chrisparton1991/micronaut-3-redis-session-scope-issue

1. Run a Redis instance on `localhost:6739`
2. Start the Micronaut app
3. Hit `GET http://localhost:8080/test` with an `Authorization: basic` header (doesn't matter what the value is)
4. Receive an HTTP 500 error, and a `java.lang.IllegalMonitorStateException`
5. In `application.yml`, set `micronaut.session.http.redis.enabled` to `false` and the error will no longer occur.

### Environment Information

- Operating System: Windows 10 Pro
- JDK version: AdoptOpenJDK 11.0.7

### Example Application

https://github.com/chrisparton1991/micronaut-3-redis-session-scope-issue

### Version

3.0.0

**Comment by graemerocher** [comment]
I can't reproduce this locally, can you provide the full stack track?

**Comment by chrisparton1991** [comment]
Yep, stacktrace is attached below.

Turns out the issue only occurs if a session cookie is provided that doesn't exist in Redis. This curl command reproduces the issue:
```
curl --location --request GET 'http://localhost:8080/test' --header 'Authorization: Basic YTph'  --cookie "SESSION=foobar"
```

I was testing in Postman with a stale session cookie and removing the cookie fixed the issue, however I'm still seeing failures in my integration tests that were passing on Micronaut 2.

Stacktrace:
```
20:44:16.213 [io-executor-thread-1] ERROR i.m.http.server.RouteExecutor - Unexpected error occurred: attempt to unlock read lock, not locked by current thread
java.lang.IllegalMonitorStateException: attempt to unlock read lock, not locked by current thread
	at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$Sync.unmatchedUnlockException(ReentrantReadWriteLock.java:448)
	at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$Sync.tryReleaseShared(ReentrantReadWriteLock.java:432)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.releaseShared(AbstractQueuedSynchronizer.java:1382)
	at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.unlock(ReentrantReadWriteLock.java:897)
	at io.micronaut.context.scope.AbstractConcurrentCustomScope.getOrCreate(AbstractConcurrentCustomScope.java:147)
	at io.micronaut.context.DefaultBeanContext.getScopedBeanForDefinition(DefaultBeanContext.java:2798)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2725)
	at io.micronaut.context.DefaultBeanContext.getProxyTargetBean(DefaultBeanContext.java:1339)
	at com.example.$RequestContext$Definition$Intercepted.$resolveTarget(Unknown Source)
	at com.example.$RequestContext$Definition$Intercepted.interceptedTarget(Unknown Source)
	at com.example.$RequestContext$Definition$Intercepted.hi(Unknown Source)
	at com.example.TestController.hi(Application.java:62)
	at com.example.$TestController$Definition$Exec.dispatch(Unknown Source)
	at io.micronaut.context.AbstractExecutableMethodsDefinition$DispatchedExecutableMethod.invoke(AbstractExecutableMethodsDefinition.java:338)
	at io.micronaut.context.DefaultBeanContext$4.invoke(DefaultBeanContext.java:503)
	at io.micronaut.web.router.AbstractRouteMatch.execute(AbstractRouteMatch.java:246)
	at io.micronaut.web.router.RouteMatch.execute(RouteMatch.java:111)
	at io.micronaut.http.server.RouteExecutor.lambda$executeRoute$15(RouteExecutor.java:605)
	at reactor.core.publisher.FluxDefer.subscribe(FluxDefer.java:46)
	at reactor.core.publisher.FluxSubscribeOn$SubscribeOnSubscriber.run(FluxSubscribeOn.java:194)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by chrisparton1991** [comment]
I can confirm the issue is resolved in 3.0.2, both in tests and regular execution. Appreciate the rapid fix!
